erts/emulator/beam/erl_nif.c

```diff
@@ -5133,6 +5223,173 @@ Eterm erts_nif_call_function(Process *p, Process *tracee,
     return nif_result;
 }
 
+/*
+ * On-halt options...
+ */
+
+int
+enif_enable_delay_halt(ErlNifEnv *env)
+{
+    struct erl_module_nif *m;
+    if (!env || !env->mod_nif)
+        return EINVAL;
+    m = env->mod_nif;
+    if (!(m->flags & ERTS_MOD_NIF_FLG_LOADING))
+        return EINVAL;
+    if (m->flags & ERTS_MOD_NIF_FLG_DELAY_HALT)
+        return EINVAL;
+    m->flags |= ERTS_MOD_NIF_FLG_DELAY_HALT;
+    return 0;
+}
+
+int
+enif_install_on_halt_handler(ErlNifEnv *env, void (*on_halt)(void *))
+{
+    ErtsNifOnHaltData *ohdp;
+    struct erl_module_nif *m;
+    if (!on_halt || !env || !env->mod_nif)
+        return EINVAL;
+    m = env->mod_nif;
+    if (!(m->flags & ERTS_MOD_NIF_FLG_LOADING) || !!m->on_halt_data)
+        return EINVAL;
+
+    ohdp = erts_alloc(ERTS_ALC_T_NIF_ON_HALT_DATA,
+                      sizeof(ErtsNifOnHaltData));
```

Comment:
Maybe not worth the refactoring, but in the name of keep-it-simple...
I guess `ErtsNifOnHaltData` could be made a part of `erl_module_nif`. It would only waste a few words per `erl_module_nif` not using `on_halt`.

Reply:
Agree. `ErtsNifOnHaltData` is now part of `struct erl_module_nif`